jinahub/indexers/searcher/HnswlibSearcher/__init__.py

```diff
@@ -0,0 +1,92 @@
+from jina_commons.indexers.dump import import_vectors
+
+
+class HnswlibSearcher(Executor):
+    """Hnswlib powered vector indexer
+
+    For more information about the Hnswlib supported parameters, please consult:
+        - https://github.com/nmslib/hnswlib
+
+    .. note::
+        Hnswlib package dependency is only required at the query time.
+    """
+
+    def __init__(
+        self,
+        default_top_k: int = 10,
+        metric: str = 'cosine',
+        dump_path: Optional[str] = None,
+        default_traversal_paths: List[str] = None,
+        **kwargs,
+    ):
+        """
+        Initialize an HnswlibSearcher
+
+        :param default_top_k: get tok k vectors
+        :param metric: Metric can be 'l2', 'ip', or 'cosine'
+        :param dump_path: the path to load ids and vecs
+        :param traverse_path: traverse path on docs, e.g. ['r'], ['c']
+        :param args:
+        :param kwargs:
+        """
+        super().__init__(**kwargs)
+        self.default_top_k = default_top_k
+        self.metric = metric
+        self.default_traversal_paths = default_traversal_paths or ['r']
+        self.logger = get_logger(self)
+        dump_path = dump_path or kwargs.get('runtime_args', {}).get('dump_path', None)
+        if dump_path is not None:
+            self.logger.info('Start building "HnswlibSearcher" from dump data')
+            ids, vecs = import_vectors(dump_path, str(self.metas.pea_id))
+            self._ids = np.array(list(ids))
+            self._vecs = np.array(list(vecs))
+            num_dim = self._vecs.shape[1]
+            self._indexer = hnswlib.Index(space=self.metric, dim=num_dim)
+            self._indexer.init_index(max_elements=len(self._vecs), ef_construction=400, M=64)
+
+            self._doc_id_to_offset = {}
+            self._load_index(self._ids, self._vecs)
+        else:
+            self.logger.warning(
+                'No data loaded in "HnswlibSearcher". Use .rolling_update() to re-initialize it...'
+            )
+
+    def _load_index(self, ids, vecs):
+        for idx, v in enumerate(vecs):
+            self._indexer.add_items(v.astype(np.float32),idx)
+            self._doc_id_to_offset[ids[idx]] = idx
+        self._indexer.set_ef(50)
+
+    @requests(on='/search')
+    def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        if not hasattr(self, '_indexer'):
+            self.logger.warning('Querying against an empty index')
+            return
+        top_k = parameters.get('top_k', self.default_top_k)
+        traversal_paths = parameters.get(
+            'traversal_paths', self.default_traversal_paths
+        )
+
+        for doc in docs.traverse_flat(traversal_paths):
+            indices, dists = self._indexer.knn_query(doc.embedding, k=top_k)
+            for idx, dist in zip(indices[0], dists[0]):
+                match = Document(id=self._ids[idx], embedding=self._vecs[idx])
+                match.scores['distance'] = 1 / (1 + dist)
```

Comment:
this computation depends on the `metric`. It only works with `cosine`?

Reply:
this needs to be clear what `dist` returns. if u say (1/1+dist) then the score is not a distance but some sense of relevance

Reply:
yes, better use `similarity` or `score`

Reply:
not only cosine, we can use 3 metrics I think

Reply:
I would use the metric name, but this needs to align with the rest of `searchers`

Reply:
I changed it to `distance`